src/test/java/org/broadinstitute/hellbender/tools/spark/transforms/markduplicates/MarkDuplicatesSparkUtilsUnitTest.java

```diff
@@ -0,0 +1,87 @@
+import org.testng.Assert;
+import org.testng.annotations.Test;
+import scala.Tuple2;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Iterator;
+import java.util.List;
+
+public class MarkDuplicatesSparkUtilsUnitTest extends BaseTest {
+    @Test
+    public void testSpanningIterator() {
+        check(Collections.emptyIterator(), Collections.emptyList());
+        check(ImmutableList.of(pair(1, "a")).iterator(),
+                ImmutableList.of(pairIterable(1, "a")));
+        check(ImmutableList.of(pair(1, "a"), pair(1, "b")).iterator(),
+                ImmutableList.of(pairIterable(1, "a", "b")));
+        check(ImmutableList.of(pair(1, "a"), pair(2, "b")).iterator(),
+                ImmutableList.of(pairIterable(1, "a"), pairIterable(2, "b")));
+        check(ImmutableList.of(pair(1, "a"), pair(1, "b"), pair(2, "c")).iterator(),
+                ImmutableList.of(pairIterable(1, "a", "b"), pairIterable(2, "c")));
+        check(ImmutableList.of(pair(1, "a"), pair(2, "b"), pair(2, "c")).iterator(),
+                ImmutableList.of(pairIterable(1, "a"), pairIterable(2, "b", "c")));
+        check(ImmutableList.of(pair(1, "a"), pair(2, "b"), pair(1, "c")).iterator(),
+                ImmutableList.of(pairIterable(1, "a"), pairIterable(2, "b"), pairIterable(1, "c")));
+    }
+
+    @Test
+    public void testSpanningIteratorWithAlternatingGroups() {
+        SAMFileHeader header = ArtificialReadUtils.createArtificialSamHeaderWithGroups(1, 1, 1000, 2);
+        GATKRead read1 = ArtificialReadUtils.createArtificialRead(header, "N", 0, 1, 20);
+        read1.setReadGroup(getReadGroupId(header, 0));
+        GATKRead read2 = ArtificialReadUtils.createArtificialRead(header, "N", 0, 2, 20);
+        read2.setReadGroup(getReadGroupId(header, 1));
+        GATKRead read3 = ArtificialReadUtils.createArtificialRead(header, "N", 0, 3, 20);
+        read3.setReadGroup(getReadGroupId(header, 0));
+        GATKRead read4 = ArtificialReadUtils.createArtificialRead(header, "N", 0, 4, 20);
+        read4.setReadGroup(getReadGroupId(header, 1));
+
+        String key1 = ReadsKey.keyForRead(header, read1);
+        String key2 = ReadsKey.keyForRead(header, read2);
+        String key3 = ReadsKey.keyForRead(header, read3);
+        String key4 = ReadsKey.keyForRead(header, read4);
+
+        Assert.assertEquals("ReadGroup0|N", key1);
+        Assert.assertEquals("ReadGroup1|N", key2);
+        Assert.assertEquals("ReadGroup0|N", key3);
+        Assert.assertEquals("ReadGroup1|N", key4);
+
+        check(ImmutableList.of(pair(key1, read1), pair(key2, read2)).iterator(),
+                ImmutableList.of(pairIterable(key1, read1), pairIterable(key2, read2)));
+        check(ImmutableList.of(pair(key1, read1), pair(key2, read2), pair(key3, read3), pair(key4, read4)).iterator(),
+                ImmutableList.of(pairIterable(key1, read1), pairIterable(key2, read2), pairIterable(key3, read3), pairIterable(key4, read4)));
```

Comment:
Is this correct behavior? In this test case, don't we want `read1` and `read3` to be grouped together (since they have the same key), and `read2` and `read4` to be similarly grouped together? It looks like you're expecting 4 groups here for only 2 keys.